google/cloud/storage/tests/curl_download_request_integration_test.cc

```diff
@@ -75,6 +77,134 @@ TEST(CurlDownloadRequestTest, SimpleStream) {
   EXPECT_EQ(kDownloadedLines, count);
 }
 
+TEST(CurlDownloadRequestTest, HandlesReleasedOnRead) {
+  auto constexpr kLineCount = 10;
+  auto constexpr kTestPoolSize = 8;
+  auto factory =
+      std::make_shared<PooledCurlHandleFactory>(kTestPoolSize, Options{});
+  ASSERT_EQ(0, factory->CurrentHandleCount());
+  ASSERT_EQ(0, factory->CurrentMultiHandleCount());
+
+  auto download = [&]() -> Status {
+    CurlRequestBuilder builder(
+        HttpBinEndpoint() + "/stream/" + std::to_string(kLineCount), factory);
+    auto download = std::move(builder).BuildDownloadRequest();
+
+    char buffer[4096];
+    auto read = download->Read(buffer, sizeof(buffer));
+    if (!read) return std::move(read).status();
+    // The data is 10 lines of about 200 bytes each, it all fits in the buffer.
+    EXPECT_LT(read->bytes_received, sizeof(buffer));
+    // This means the transfers completes during the Read() call, and the
+    // handles are immediately returned to the pool.
+    EXPECT_EQ(1, factory->CurrentHandleCount());
+    EXPECT_EQ(1, factory->CurrentMultiHandleCount());
+
+    auto close = download->Close();
+    if (!close) return std::move(close).status();
+    EXPECT_EQ(1, factory->CurrentHandleCount());
+    EXPECT_EQ(1, factory->CurrentMultiHandleCount());
+    return Status{};
+  };
+
+  auto delay = std::chrono::seconds(1);
+  Status status;
+  for (int i = 0; i != 3; ++i) {
+    status = download();
+    if (status.ok()) break;
+    std::this_thread::sleep_for(delay);
+    delay *= 2;
+  }
+  EXPECT_EQ(1, factory->CurrentHandleCount());
+  EXPECT_EQ(1, factory->CurrentMultiHandleCount());
+  ASSERT_STATUS_OK(status);
```

Comment:
Should this happen before the handle-count expectations?

Reply:
Added a comment. It should work even if `ASSERT_STATUS_OK()` fails.  Well, at least for now, I have another PR that will change this.